### DISCUSSION TITLE
Skill rust code review

### DISCUSSION THREAD

**Discussion by Foxtrot-Uniform42** [CleverRaven/Cataclysm-DDA#46789] [General]
Recently I've noticed that higher intelligence increases **skill rust** value and vice versa (same behavior applies for `FORGETFUL` and `GOODMEMORY` traits):
![image](https://user-images.githubusercontent.com/46412513/104806519-e8415100-57f9-11eb-99ab-cbd70b8b92da.png)

It seemed strange to me, that increasing intelligence leads to incrementing skill rust value, so I created an issue: #46598
Shortly after, a fix #46605 was applied by swaping `intel - 8` values in code below:
`int ret = ( ( rate_option == "vanilla" || rate_option == "capped" ) ?
                100 : 100 + 10 * ( intel - 8 ) );`

I got curious why skill rust values was inverted in a first place, and found PR's which supposedly change what `rust_rate()` means.
The changes in skill rust system were merged to master by #37281, and changes to traits by #37883.
Authors of these PR's stated, that:
> In new skill rust system increasing skillrust rate increase time between skill rusting. Old good memory traits decrease time (accelerate skillrusting instead of slowing down) in new system.

Currently we got '**old**' skill rust system intelligence stat (e.g. higher intelligence leads to **lower** skill rust), and traits with '**new**' skill rust system (e.g. `GOODMEMORY` **increases** skill rust value). As you see, the current state of skill rust system is contradictory, and needs conclusion.

So I requesting help in reviewing skill rust code, to figure out what **actually** it is doing, to finish with this issue:
- If it is indeed increasing delay between rusts, then we should revert changes by #46605, and change naming for skill rust parameter to better indication on what it's doing.
- If it is not, then we should invert `GOODMEMORY` and `FORGETFUL` traits to finalise solution by #46605

Upd: Testing showed, that higher skill rust value leads to slower rusting in skills and vice versa. However, if the player has skill rust below 1% (low intelligence + forgetful trait), then it is **completely turned off**, which I think is a main problem with this new system.
Upd 2: Skill rust was renamed to skill rust delay

**Comment by actual-nh** [comment]
If a skill is disabled for learning+focus loss, does using it actually prevent skill rust? The check in `character::practice()` for returning if a skill is disabled is prior to calling `SkillLevel::practice()`.

**Reply by actual-nh** [reply]
@Foxtrot-Uniform42: You're experienced in testing skill rust; could you check on the above?

**Reply by Foxtrot-Uniform42** [reply]
I gone to swim with disabled athletics, and it prevented me from skill rusting, so yes, **practice disabled skills prevents it from rust**.

**Reply by actual-nh** [reply]
Weird. (I have to wonder if disabled skills rust at all...) Thanks!

**Reply by Foxtrot-Uniform42** [reply]
Yes, disabled skills rusts if left unpracticed, I checked that too.

**Reply by Foxtrot-Uniform42** [reply]
I kinda like current behavior with disabled skills counts as practice for rust rate. When you not spending your focus at something to learn new things, at least you should be able to maintain your proficiency at what you already know.

**Comment by eltank** [comment]
The way skill rust works was overhauled in #50176.
The code being referenced in this discussion (Character::rust_rate) is about to be removed in #51327.
I think that should resolve (or obsolete) the issues in this topic.

**Comment by deane-c64** [comment]
related, but not totally on topic. is skill rust Supposed to be that you start forgetting after 1 day? I don't have good or bad memory. and HIGH as in 20+ intelligence. it bugs me that after 1 day, that I start to lose levels to rust.

if not, please try to find out why it is; can give video of it. if so, still thanks for reading. sorry if I've revived dead topic. and please have a wonder rest of the weekend.
